diff --git a/server/src/main/java/org/gluu/oxtrust/action/Authenticator.java b/server/src/main/java/org/gluu/oxtrust/action/Authenticator.java
index 1c1fa2b..4caed76 100644
--- a/server/src/main/java/org/gluu/oxtrust/action/Authenticator.java
+++ b/server/src/main/java/org/gluu/oxtrust/action/Authenticator.java
@@ -115,7 +115,7 @@ public class Authenticator implements Serializable {
 	String viewIdBeforeLoginRedirect;
 
 	@In(create = true)
-	@Out(scope = ScopeType.SESSION)
+	@Out(scope = ScopeType.SESSION, required = false)
 	private OauthData oauthData;
 
 	@In(value = "#{oxTrustConfiguration.applicationConfiguration}")
@@ -281,9 +281,15 @@ public class Authenticator implements Serializable {
 	public void processLogout() throws Exception {
 		ssoLoginAction.logout();
 		oAuthlLogout();
+
+		postLogout();
 	}
 
 	public String postLogout() {
+		if (identity.isLoggedIn()) {
+			identity.logout();
+		}
+
 		return OxTrustConstants.RESULT_SUCCESS;
 	}
 
@@ -302,6 +308,7 @@ public class Authenticator implements Serializable {
 		oauthData.setUserUid(null);
 		oauthData.setIdToken(null);
 		oauthData.setSessionId(null);
+		oauthData = null;
 
 		FacesContext.getCurrentInstance().getExternalContext().redirect(clientRequest.getUri());
 	}
diff --git a/server/src/main/java/org/gluu/oxtrust/action/SsoLoginAction.java b/server/src/main/java/org/gluu/oxtrust/action/SsoLoginAction.java
index 118c083..90787b2 100644
--- a/server/src/main/java/org/gluu/oxtrust/action/SsoLoginAction.java
+++ b/server/src/main/java/org/gluu/oxtrust/action/SsoLoginAction.java
@@ -173,10 +173,6 @@ public class SsoLoginAction implements Serializable {
 			}
 		}
 
-		if (identity.isLoggedIn()) {
-			identity.logout();
-		}
-
 		return isShib2Authentication ? OxTrustConstants.RESULT_LOGOUT_SSO : OxTrustConstants.RESULT_LOGOUT;
 	}
 
